File: go/v4/exchange_sbe.go
package ccxt

import (
	"fmt"
	"strings"
)

// ============================================================================
// SBE (Simple Binary Encoding) Helper Methods
// ============================================================================
// This file contains SBE-specific helper methods for the Exchange class.
// These methods provide utilities for WebSocket binary message decoding,
// orderbook formatting, and data normalization for SBE-encoded messages.
// ============================================================================

// SetupSbeBinaryDecoder sets up binary decoder for SBE WebSocket connections
// Check if URL indicates SBE format and decoder not already set
func (this *Exchange) SetupSbeBinaryDecoder(client interface{}, url interface{}, decoderFunction interface{}) {
	urlStr, ok := url.(string)
	if !ok {
		return
	}

	isSbeUrl := strings.Contains(urlStr, "responseFormat=sbe") ||
		strings.Contains(urlStr, "/public-sbe") ||
		strings.Contains(urlStr, "/ws-sbe")

	if isSbeUrl && client != nil {
		// In Go, client properties are accessed through reflection or type assertion
		// The actual implementation depends on the WS client structure
		// For now, we just check if SBE URL and leave decoder setup to specific exchange implementations
		if this.Verbose {
			fmt.Println("SetupSbeBinaryDecoder: Detected SBE URL:", urlStr)
		}
	}
}

// DecodeSbeNestedMessage decodes nested SBE message (used in envelope patterns like Binance WebSocket)
// @param buffer - The binary SBE message
// @returns Decoded message data
func (this *Exchange) DecodeSbeNestedMessage(buffer interface{}) interface{} {
	templateId := this.ReadSbeTemplateId(buffer)
	if this.Verbose {
		var bufLen int
		if bytes, ok := buffer.([]byte); ok {
			bufLen = len(bytes)
		}
		fmt.Println("DecodeSbeNestedMessage: templateId:", templateId, "buffer length:", bufLen)
	}

	registry := this.GetSbeDecoderRegistry()
	if registry == nil {
		if this.Verbose {
			fmt.Println("DecodeSbeNestedMessage: no decoder registry available")
		}
		return nil
	}

	registryMap, ok := registry.(map[string]interface{})
	if !ok {
		return nil
	}

	templateIdStr := fmt.Sprintf("%v", templateId)
	_, exists := registryMap[templateIdStr]
	if !exists {
		if this.Verbose {
			fmt.Println("DecodeSbeNestedMessage: unknown template ID:", templateId)
		}
		return nil
	}

	// Defer error handling
	defer func() {
		if r := recover(); r != nil {
			if this.Verbose {
				fmt.Println("DecodeSbeNestedMessage: error decoding templateId", templateId, ":", r)
			}
		}
	}()

	// Call the decoder - implementation depends on decoder type
	result := this.DecodeSbeMessage(buffer, registry)
	if resultMap, ok := result.(map[string]interface{}); ok {
		return resultMap["data"]
	}
	return result
}

// DecodeSbeWebSocketMessage decodes SBE-encoded WebSocket message
// Override this in exchange class to handle exchange-specific envelope patterns
// @param buffer - The binary SBE message
// @returns Decoded message data
func (this *Exchange) DecodeSbeWebSocketMessage(buffer interface{}) interface{} {
	// Default implementation: direct decode without envelope
	result := this.DecodeSbeMessage(buffer, this.GetSbeDecoderRegistry())
	if resultMap, ok := result.(map[string]interface{}); ok {
		return resultMap["data"]
	}
	return result
}

// GetSbeMessageHandlers gets SBE message handlers registry
// Override this in exchange class to provide message routing
// @returns Map of template ID to handler function
func (this *Exchange) GetSbeMessageHandlers() interface{} {
	return map[string]interface{}{}
}

// HandleSbeMessage handles SBE binary message with automatic routing
// @param client - WebSocket client
// @param message - Binary SBE message
func (this *Exchange) HandleSbeMessage(client interface{}, message interface{}) {
	defer func() {
		if r := recover(); r != nil {
			if this.Verbose {
				fmt.Println("HandleSbeMessage: Failed to decode SBE message:", r)
			}
		}
	}()

	if this.Verbose {
		var msgLen int
		if bytes, ok := message.([]byte); ok {
			msgLen = len(bytes)
		}
		fmt.Println("HandleSbeMessage: Received binary message, size:", msgLen)
	}

	// Decode the SBE message
	decoderRegistry := this.GetSbeDecoderRegistry()
	result := this.DecodeSbeMessage(message, decoderRegistry)

	resultMap, ok := result.(map[string]interface{})
	if !ok {
		return
	}

	templateId := resultMap["templateId"]
	decoded := resultMap["data"]

	if this.Verbose {
		fmt.Println("HandleSbeMessage: Decoded template ID:", templateId)
	}

	// Route to appropriate handler
	handlers := this.GetSbeMessageHandlers()
	handlersMap, ok := handlers.(map[string]interface{})
	if !ok {
		return
	}

	templateIdStr := fmt.Sprintf("%v", templateId)
	handler, exists := handlersMap[templateIdStr]
	if exists {
		if handlerFunc, ok := handler.(func(interface{}, interface{})); ok {
			handlerFunc(client, decoded)
		}
	} else if this.Verbose {
		fmt.Println("HandleSbeMessage: No handler for template ID:", templateId)
	}
}

// FormatSbeOrderbook formats orderbook from SBE message with exponents
// Convenience helper for common orderbook processing pattern
// @param message - Decoded SBE message
// @param bidsKey - Key for bids array (default: "bids")
// @param asksKey - Key for asks array (default: "asks")
// @param priceKey - Key for price in bid/ask objects (default: "px")
// @param sizeKey - Key for size in bid/ask objects (default: "sz")
// @returns Formatted orderbook with bids and asks arrays
func (this *Exchange) FormatSbeOrderbook(message interface{}, keys ...string) interface{} {
	bidsKey := "bids"
	asksKey := "asks"
	priceKey := "px"
	sizeKey := "sz"

	if len(keys) > 0 {
		bidsKey = keys[0]
	}
	if len(keys) > 1 {
		asksKey = keys[1]
	}
	if len(keys) > 2 {
		priceKey = keys[2]
	}
	if len(keys) > 3 {
		sizeKey = keys[3]
	}

	messageMap, ok := message.(map[string]interface{})
	if !ok {
		return map[string]interface{}{
			"bids": []interface{}{},
			"asks": []interface{}{},
		}
	}

	pxExponent := this.SafeInteger(message, "pxExponent", this.SafeInteger(message, "priceExponent", 0))
	szExponent := this.SafeInteger(message, "szExponent", this.SafeInteger(message, "qtyExponent", 0))

	rawBids := this.SafeValue(messageMap, bidsKey, []interface{}{})
	rawAsks := this.SafeValue(messageMap, asksKey, []interface{}{})

	rawBidsList, _ := rawBids.([]interface{})
	rawAsksList, _ := rawAsks.([]interface{})

	bids := []interface{}{}
	asks := []interface{}{}

	// Process bids
	for _, bidItem := range rawBidsList {
		bid, ok := bidItem.(map[string]interface{})
		if !ok {
			continue
		}
		pxMantissa := this.SafeInteger(bid, priceKey+"Mantissa", this.SafeInteger(bid, priceKey, 0))
		szMantissa := this.SafeInteger(bid, sizeKey+"Mantissa", this.SafeInteger(bid, sizeKey, 0))
		price := this.ApplyExponent(pxMantissa, pxExponent)
		size := this.ApplyExponent(szMantissa, szExponent)
		bids = append(bids, []interface{}{price, size})
	}

	// Process asks
	for _, askItem := range rawAsksList {
		ask, ok := askItem.(map[string]interface{})
		if !ok {
			continue
		}
		pxMantissa := this.SafeInteger(ask, priceKey+"Mantissa", this.SafeInteger(ask, priceKey, 0))
		szMantissa := this.SafeInteger(ask, sizeKey+"Mantissa", this.SafeInteger(ask, sizeKey, 0))
		price := this.ApplyExponent(pxMantissa, pxExponent)
		size := this.ApplyExponent(szMantissa, szExponent)
		asks = append(asks, []interface{}{price, size})
	}

	return map[string]interface{}{
		"bids": bids,
		"asks": asks,
	}
}

// SafeBigInt safely accesses a BigInt value and converts to number
// Many SBE decoders return BigInt for large numbers - this helper safely converts them
// @param object - The object containing the value
// @param key - The key to access
// @param defaultValue - Default value if key not found
// @returns The value as number (converted from BigInt if needed)
func (this *Exchange) SafeBigInt(object interface{}, key interface{}, defaultValue ...interface{}) interface{} {
	value := this.SafeValue(object, key)
	if value == nil {
		if len(defaultValue) > 0 {
			return defaultValue[0]
		}
		return nil
	}

	// In Go, handle int64, uint64 which are similar to BigInt
	switch v := value.(type) {
	case int64:
		return float64(v)
	case uint64:
		return float64(v)
	case int:
		return float64(v)
	case float64:
		return v
	default:
		return value
	}
}

// ConvertBigIntFields converts all BigInt fields in an object to numbers
// Useful for SBE messages that use BigInt for IDs and timestamps
// @param object - The object to process
// @param fields - Specific fields to convert (optional)
// @returns New object with BigInt fields converted to numbers
func (this *Exchange) ConvertBigIntFields(object interface{}, fields ...interface{}) interface{} {
	if object == nil {
		return object
	}

	objectMap, ok := object.(map[string]interface{})
	if !ok {
		return object
	}

	result := make(map[string]interface{})

	var keys []string
	if len(fields) > 0 {
		if fieldsList, ok := fields[0].([]string); ok {
			keys = fieldsList
		} else {
			keys = make([]string, 0, len(objectMap))
			for k := range objectMap {
				keys = append(keys, k)
			}
		}
	} else {
		keys = make([]string, 0, len(objectMap))
		for k := range objectMap {
			keys = append(keys, k)
		}
	}

	for _, key := range keys {
		value := objectMap[key]

		switch v := value.(type) {
		case int64:
			result[key] = float64(v)
		case uint64:
			result[key] = float64(v)
		case []interface{}:
			converted := make([]interface{}, len(v))
			for j, item := range v {
				converted[j] = this.ConvertBigIntFields(item, fields...)
			}
			result[key] = converted
		default:
			result[key] = value
		}
	}

	return result
}

// NormalizeSbeOrder normalizes SBE order data (common pattern for Binance WebSocket)
// Converts exponent fields to actual values
// @param order - SBE order data
// @returns Normalized order data compatible with parseOrder
func (this *Exchange) NormalizeSbeOrder(order interface{}) interface{} {
	orderMap, ok := order.(map[string]interface{})
	if !ok {
		return order
	}

	// Check if this is SBE format (has exponent fields)
	_, hasPriceExp := orderMap["priceExponent"]
	_, hasQtyExp := orderMap["qtyExponent"]
	hasExponents := hasPriceExp || hasQtyExp

	if !hasExponents {
		return order // Already normalized or not SBE format
	}

	// Apply exponents and convert to standard format
	priceExponent := this.SafeInteger(order, "priceExponent", 0)
	qtyExponent := this.SafeInteger(order, "qtyExponent", 0)

	normalized := make(map[string]interface{})

	for key, value := range orderMap {
		// Skip exponent fields themselves
		if strings.HasSuffix(key, "Exponent") {
			continue
		}

		// Handle price-related mantissa fields
		if key == "price" || key == "stopPrice" || key == "icebergQty" {
			normalized[key] = this.NumberToString(this.ApplyExponent(value, priceExponent))
		} else if key == "origQty" || key == "executedQty" || key == "cummulativeQuoteQty" {
			// Handle quantity-related fields
			normalized[key] = this.NumberToString(this.ApplyExponent(value, qtyExponent))
		} else {
			// Convert int64/uint64 to string for IDs
			switch v := value.(type) {
			case int64:
				normalized[key] = fmt.Sprintf("%d", v)
			case uint64:
				normalized[key] = fmt.Sprintf("%d", v)
			default:
				// Copy other fields as-is
				normalized[key] = value
			}
		}
	}

	return normalized
}

File: cs/ccxt/base/Exchange.Sbe.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ccxt;

using dict = Dictionary<string, object>;

/// <summary>
/// SBE (Simple Binary Encoding) related methods for decoding binary market data
/// </summary>
public partial class Exchange
{
    /// <summary>
    /// Reads the template ID from an SBE message header
    /// SBE message header format (8 bytes):
    ///   - blockLength (2 bytes, uint16, offset 0)
    ///   - templateId (2 bytes, uint16, offset 2)
    ///   - schemaId (2 bytes, uint16, offset 4)
    ///   - version (2 bytes, uint16, offset 6)
    /// </summary>
    /// <param name="buffer">The binary SBE message (byte array or string)</param>
    /// <returns>Template ID as integer</returns>
    public object readSbeTemplateId(object buffer)
    {
        byte[] bufferBytes;

        // Convert to byte array
        if (buffer is byte[] bytes)
        {
            bufferBytes = bytes;
        }
        else if (buffer is string str)
        {
            bufferBytes = Encoding.UTF8.GetBytes(str);
        }
        else
        {
            return -1;
        }

        if (bufferBytes.Length < 8)
        {
            return -1;
        }

        // Read uint16 at offset 2, little-endian
        int templateId = BitConverter.ToUInt16(bufferBytes, 2);
        return templateId;
    }

    /// <summary>
    /// Convert mantissa and exponent to decimal number
    /// Used by exchanges that encode prices/amounts as mantissa * 10^exponent
    /// </summary>
    /// <param name="mantissa">The mantissa value (int, long, or double)</param>
    /// <param name="exponent">The power of 10 exponent</param>
    /// <returns>The decimal number</returns>
    public object applyExponent(object mantissa, object exponent)
    {
        double mantissaNum;
        double exponentNum;

        // Convert mantissa to double
        if (mantissa is int || mantissa is long || mantissa is short || mantissa is byte)
        {
            mantissaNum = Convert.ToDouble(mantissa);
        }
        else if (mantissa is double || mantissa is float || mantissa is decimal)
        {
            mantissaNum = Convert.ToDouble(mantissa);
        }
        else if (mantissa is string mantissaStr)
        {
            if (!double.TryParse(mantissaStr, out mantissaNum))
            {
                mantissaNum = 0.0;
            }
        }
        else
        {
            mantissaNum = 0.0;
        }

        // Convert exponent to double
        if (exponent is int || exponent is long || exponent is short || exponent is byte)
        {
            exponentNum = Convert.ToDouble(exponent);
        }
        else if (exponent is double || exponent is float || exponent is decimal)
        {
            exponentNum = Convert.ToDouble(exponent);
        }
        else if (exponent is string exponentStr)
        {
            if (!double.TryParse(exponentStr, out exponentNum))
            {
                exponentNum = 0.0;
            }
        }
        else
        {
            exponentNum = 0.0;
        }

        return mantissaNum * Math.Pow(10, exponentNum);
    }

    /// <summary>
    /// Convert mantissa128 byte array to number
    /// mantissa128 is a signed 128-bit integer encoded as a little-endian byte array
    /// Used by Binance SBE for large volume values
    /// </summary>
    /// <param name="bytesData">The byte array</param>
    /// <returns>The numeric value</returns>
    public object mantissa128ToNumber(object bytesData)
    {
        if (bytesData == null)
        {
            return 0;
        }

        byte[] bytes;

        // Convert to byte array
        if (bytesData is byte[] byteArray)
        {
            if (byteArray.Length == 0)
            {
                return 0;
            }
            bytes = byteArray;
        }
        else if (bytesData is string str)
        {
            if (str.Length == 0)
            {
                return 0;
            }
            bytes = Encoding.UTF8.GetBytes(str);
        }
        else if (bytesData is List<object> list)
        {
            if (list.Count == 0)
            {
                return 0;
            }
            bytes = new byte[list.Count];
            for (int i = 0; i < list.Count; i++)
            {
                bytes[i] = Convert.ToByte(list[i]);
            }
        }
        else
        {
            return 0;
        }

        // For mantissa128, we need to handle signed 128-bit integers
        // The value −2^127 is nullValue by default
        // For practical purposes, we'll read up to 8 bytes (64-bit) as most values fit
        ulong result = 0;
        ulong multiplier = 1;

        // Read up to 8 bytes (64-bit safe range)
        int limit = Math.Min(bytes.Length, 8);

        for (int i = 0; i < limit; i++)
        {
            result += (ulong)bytes[i] * multiplier;
            multiplier *= 256;
        }

        return (long)result;
    }

    /// <summary>
    /// Decode SBE message using registered decoder based on template ID
    /// </summary>
    /// <param name="buffer">The binary SBE message (byte array or string)</param>
    /// <param name="decoderRegistry">Map of template ID to decoder class</param>
    /// <returns>Decoded message with templateId and data</returns>
    public object decodeSbeMessage(object buffer, object decoderRegistry)
    {
        byte[] bufferBytes;

        // Convert to byte array
        if (buffer is byte[] bytes)
        {
            bufferBytes = bytes;
        }
        else if (buffer is string str)
        {
            bufferBytes = Encoding.UTF8.GetBytes(str);
        }
        else
        {
            throw new ExchangeError(this.id + " decodeSbeMessage() invalid buffer type");
        }

        // Read template ID from header
        var templateId = readSbeTemplateId(bufferBytes);

        // Look up decoder class in registry
        if (!(decoderRegistry is dict registry))
        {
            throw new ExchangeError(this.id + " decodeSbeMessage() invalid decoder registry");
        }

        string templateIdStr = templateId.ToString();
        if (!registry.ContainsKey(templateIdStr))
        {
            throw new ExchangeError(this.id + " decodeSbeMessage() unknown template ID: " + templateIdStr);
        }

        var decoderClass = registry[templateIdStr];

        // Instantiate and decode
        // Skip 8-byte SBE message header (blockLength, templateId, schemaId, version)
        // Note: In C#, the decoder implementation would need to be called here
        // This is a placeholder that assumes the decoder has a Decode method
        dynamic decoder = Activator.CreateInstance(decoderClass.GetType());
        var decoded = decoder.decode(bufferBytes, 8);

        return new dict
        {
            ["templateId"] = templateId,
            ["data"] = decoded
        };
    }

    /// <summary>
    /// Automatically apply exponents to mantissa fields in decoded SBE data
    /// Finds fields ending in 'Mantissa' and applies corresponding 'Exponent' field
    /// </summary>
    /// <param name="data">Decoded SBE message data</param>
    /// <param name="exponentMap">Optional map of mantissa field names to exponent field names</param>
    /// <returns>Data with mantissas converted to decimal numbers</returns>
    public object applySbeExponents(object data, object exponentMap = null)
    {
        if (data == null || !(data is dict dataMap))
        {
            return data;
        }

        var result = new dict();
        var expMap = exponentMap as dict;

        // Get all exponent fields first
        var exponents = new dict();
        foreach (var kvp in dataMap)
        {
            if (kvp.Key.EndsWith("Exponent"))
            {
                exponents[kvp.Key] = kvp.Value;
            }
        }

        // Process all fields
        foreach (var kvp in dataMap)
        {
            string key = kvp.Key;
            object value = kvp.Value;

            // Handle arrays recursively
            if (value is List<object> list)
            {
                var mappedArray = new List<object>();
                foreach (var item in list)
                {
                    mappedArray.Add(applySbeExponents(item, exponentMap));
                }
                result[key] = mappedArray;
            }
            else if (key.EndsWith("Mantissa"))
            {
                // Find corresponding exponent
                string baseFieldName = key.Replace("Mantissa", "");
                string exponentKey = baseFieldName + "Exponent";

                // Check if custom exponent mapping provided
                if (expMap != null && expMap.ContainsKey(key))
                {
                    exponentKey = Convert.ToString(expMap[key]);
                }

                object exponent = exponents.ContainsKey(exponentKey) ? exponents[exponentKey] : 0;

                // Convert mantissa to decimal number
                result[baseFieldName] = applyExponent(value, exponent);
            }
            else if (!key.EndsWith("Exponent"))
            {
                // Copy non-exponent fields
                result[key] = value;
            }
        }

        return result;
    }

    /// <summary>
    /// Decode SBE-encoded WebSocket message
    /// Override this in exchange class to handle exchange-specific envelope patterns
    /// </summary>
    /// <param name="buffer">The binary SBE message</param>
    /// <returns>Decoded message data</returns>
    public virtual object decodeSbeWebSocketMessage(object buffer)
    {
        // Default implementation: direct decode without envelope
        object decoderRegistry = this.call("getSbeDecoderRegistry");
        object result = this.decodeSbeMessage(buffer, decoderRegistry);
        return ((dict)result)["data"];
    }

    /// <summary>
    /// Get SBE message handlers registry
    /// Override this in exchange class to provide message routing
    /// </summary>
    /// <returns>Map of template ID to handler function</returns>
    public virtual object getSbeMessageHandlers()
    {
        return new dict();
    }

    /// <summary>
    /// Normalize SBE order format to standard order format
    /// </summary>
    /// <param name="order">SBE order data</param>
    /// <returns>Normalized order data compatible with parseOrder</returns>
    public virtual object normalizeSbeOrder(object order)
    {
        // Check if this is SBE format (has exponent fields)
        dict orderDict = order as dict;
        if (orderDict == null)
        {
            return order;
        }

        bool hasExponents = orderDict.ContainsKey("priceExponent") || orderDict.ContainsKey("qtyExponent");
        if (!hasExponents)
        {
            return order; // Already normalized or not SBE format
        }

        // Apply exponents and convert to standard format
        return this.applySbeExponents(order, null);
    }
}
